src/haproxy/conn.go

```diff
@@ -0,0 +1,126 @@
+package haproxy
+
+import (
+	"bufio"
+	"net"
+	"time"
+
+	proxyproto "github.com/pires/go-proxyproto"
+)
+
+// Conn is a `net.Conn` compatible struct that handles PROXY header checking.
+type Conn struct {
+	rd  *bufio.Reader
+	c   net.Conn
+	r   net.Addr
+	l   net.Addr
+	hdr *proxyproto.Header
+}
+
+// NewConn wraps a connection with a `bufio.Reader` for checking if the
+// PROXY protocol headers are present
+func NewConn(nc net.Conn) (net.Conn, error) {
+	c := &Conn{
+		c:  nc,
+		rd: bufio.NewReader(nc),
+	}
+	if err := c.ProxyInit(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// ProxyInit checks for the PROXY protocol headers, and if not returns
+// the connection unmolested.
+func (c *Conn) ProxyInit() error {
+	pc, err := proxyproto.Read(c.rd)
+	if err != nil &&
+		err != proxyproto.ErrNoProxyProtocol &&
+		err != proxyproto.ErrInvalidLength {
+		return err
+	} else if err == proxyproto.ErrInvalidLength {
+		// Failed to parse header, the connection is probably about to drop anyway,
+		// it's not a PROXY header at least.
+		return nil
+	}
+
+	c.hdr = pc
+	return nil
+}
+
+// Read reads data from the connection.
+// Read can be made to time out and return an Error with Timeout() == true
+// after a fixed time limit; see SetDeadline and SetReadDeadline.
```

Comment:
Not sure what `Timeout()` refers to here, probs a copy/paste?

Reply:
It is relevant, but also copy/paste.

net.Error has a Timeout() method, I've corrected the comment to refer to net.Error (since the original documentation `Error` would refer to `net.Error`)